File: apps/frontend/types/order.ts
export type OrderStatus =
  | "PENDING"
  | "CONFIRMED"
  | "SHIPPED"
  | "DELIVERED"
  | "CANCELLED";

export type OrderCreatePayload = {
  customer_name: string;
  phone_number: string;
  address: string;
};

export type OrderItem = {
  id: number;
  order_id: number;
  product_id: number;
  quantity: number;
  price: string | number;
  subtotal: string | number;
};

export type Order = {
  id: number;
  user_id: number;
  status: OrderStatus;
  total_price: string | number;
  customer_name: string;
  phone_number: string;
  address: string;
  created_at: string;
  updated_at: string;
  items: OrderItem[];
};

File: apps/backend/app/schemas/order.py
from datetime import datetime
from decimal import Decimal
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.common import ListMeta


class OrderStatus(str, Enum):
    PENDING = "PENDING"
    CONFIRMED = "CONFIRMED"
    SHIPPED = "SHIPPED"
    DELIVERED = "DELIVERED"
    CANCELLED = "CANCELLED"


class OrderCreate(BaseModel):
    customer_name: str = Field(min_length=2, max_length=255)
    phone_number: str = Field(min_length=5, max_length=40)
    address: str = Field(min_length=5, max_length=1000)

    @field_validator("customer_name", "phone_number", "address")
    @classmethod
    def normalize_text(cls, value: str) -> str:
        return " ".join(value.strip().split())


class OrderStatusUpdate(BaseModel):
    status: OrderStatus


class OrderItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    order_id: int
    product_id: int
    quantity: int
    price: Decimal
    subtotal: Decimal


class OrderResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    status: OrderStatus
    total_price: Decimal
    customer_name: str
    phone_number: str
    address: str
    created_at: datetime
    updated_at: datetime
    items: list[OrderItemResponse] = []


class OrderListResponse(BaseModel):
    items: list[OrderResponse]
    meta: ListMeta
